### ISSUE TITLE
c++ backend thread pool overhead

### ISSUE THREAD

**Issue by newpoo** [tensorflow/tensorflow#25637]
Is there a configuration to eliminate the intra and inter thread pool and always do computation in C++ client threads that call session->run().

**Comment by jvishnuvardhan** [comment]
@newpoo Could you describe more details about the approach and features you are looking? Thanks!

**Comment by newpoo** [comment]
@jvishnuvardhan If possible, essentially we'd like to eliminate the two thread pools (inter_op and intra_op pools) inside tensorflow library , and run session->run() in client thread. Because we run online inference on CPU, and we'd like to reduce the context switch introduced by these two thread pools.

**Comment by alextp** [comment]
You can try to set the thread pool sizes to 1, but this will only buy you so much. The intra-op pool is relatively harmless, but the inter-op pool is necessary to avoid deadlocks in some cases. Note also that TF has more thread pools, like the event manager pool.

Instead I recommend you look at using tf lite, which doesn't use threads by default.

**Comment by sjoshi6** [comment]
We were facing a similar issue where context switch from service threads to inter_op threads was very high. 

Found the `set_inter_op_thread_pool` in `tensorflow::RunOptions` on line https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/protobuf/config.proto#L632 

When this flag was set to `RunOptions.set_inter_op_thread_pool(-1)` before calling the `DirectSession::Run` we were able to bypass the inter-op thread-pool during graph execution. 

The graph execution was driven by the caller thread as expected, which helped us resolve this issue. 

Although this still creates the inter-op thread pool, we were able disable it's effect by setting the pool size to 1 and bypassing it using the `RunOptions.set_inter_op_thread_pool(-1)`.
